session.go

```diff
@@ -2484,7 +2484,15 @@ func (c *Collection) Upsert(selector interface{}, update interface{}) (info *Cha
 		Flags:      1,
 		Upsert:     true,
 	}
-	lerr, err := c.writeOp(&op, true)
+	var lerr *LastError
+	for {
```

Comment:
Let's please start by retrying at most 5 times here and in the other loop. More than that there's a good chance that there's something else wrong.